src/ffi/CXXInterface.h

```diff
@@ -0,0 +1,645 @@
+        const_cast<clang::NamedDecl*>(decl))};
+      return ast->getTemplateSpecializationType(templ, args);
+    }
+
+    /**
+     * Get the canonical template especialization with args.
+     */
+    clang::QualType getCanonicalTemplateSpecializationType(
+      const clang::NamedDecl* decl,
+      llvm::ArrayRef<clang::TemplateArgument> args)
+    {
+      clang::TemplateName templ{clang::dyn_cast<clang::TemplateDecl>(
+        const_cast<clang::NamedDecl*>(decl))};
+      return ast->getCanonicalTemplateSpecializationType(templ, args);
+    }
+
+    /**
+     * Instantiate a new function at the end of the main file, if not yet done.
+     */
+    clang::FunctionDecl* instantiateFunction(
+      const char* name, llvm::ArrayRef<CXXType> args, CXXType ret)
+    {
+      auto* DC = ast->getTranslationUnitDecl();
+      clang::SourceLocation loc = Clang->getEndOfFileLocation();
+      clang::IdentifierInfo& fnNameIdent = ast->Idents.get(name);
+      clang::DeclarationName fnName{&fnNameIdent};
+      clang::FunctionProtoType::ExtProtoInfo EPI;
+
+      // Get type of args/ret, function
+      llvm::SmallVector<clang::QualType> argTys;
+      for (auto argTy : args)
+        argTys.push_back(getQualType(argTy));
+      auto retTy = getQualType(ret);
+      clang::QualType fnTy = ast->getFunctionType(retTy, argTys, EPI);
+
+      // Create a new function
+      auto func = clang::FunctionDecl::Create(
+        *ast,
+        DC,
+        loc,
+        loc,
+        fnName,
+        fnTy,
+        ast->getTrivialTypeSourceInfo(fnTy),
+        clang::StorageClass::SC_None);
+
+      // Associate with the translation unit
+      func->setLexicalDeclContext(DC);
+      DC->addDecl(func);
+
+      return func;
+    }
+
+    /**
+     * Create a function argument
+     *
+     * FIXME: Do we want to have this as part of instantiateFunction?
+     */
+    clang::ParmVarDecl* createFunctionArgument(
+      const char* name, CXXType& ty, clang::FunctionDecl* func)
+    {
+      clang::SourceLocation loc = func->getLocation();
+      clang::IdentifierInfo& ident = ast->Idents.get(name);
+      clang::ParmVarDecl* arg = clang::ParmVarDecl::Create(
+        *ast,
+        func,
+        loc,
+        loc,
+        &ident,
+        getQualType(ty),
+        nullptr,
+        clang::StorageClass::SC_None,
+        nullptr);
+      func->setParams({arg});
+      return arg;
+    }
+
+    /**
+     * Create integer constant literal
+     *
+     * TODO: Can we have a generic literal creator or do we need one each?
+     */
+    clang::IntegerLiteral*
+    createIntegerLiteral(unsigned int len, unsigned long val)
+    {
+      llvm::APInt num{len, val};
+      auto* lit = clang::IntegerLiteral::Create(
+        *ast, num, getQualType(CXXType::getInt()), clang::SourceLocation{});
+      return lit;
+    }
+
+    /**
+     * Create a return instruction
+     *
+     * TODO: Can we have a generic instruction creator or do we need one each?
```

Comment:
We need very few things here.  The point of this function and the related ones are to generate functions that hide all of the C++ calling convention logic, like:
```c++
// Everything in this is a primitive type or pointer, with trivial alignment and layout requirements:
struct vector_int_pop_argframe
{
  int ret;
  std::vector<int> *v;
};
void call_std_vector_int_pop(vector_pop_argframe *args) {
  args->ret = args->v->pop_back();
}
```

We'll then generate a wrapper, so if we do something like this in verona:

```verona
try
{
  var i = someSandbox.vec.pop_back();
}
catch (oops: SandboxExploded)
{
}
```

On the Verona side, we'll generate a struct containing the vtable index of the exported function, a pointer to the vector, and space for the result and pass it into the sandbox.  On the sandbox side, we'll generate an LLVM function roughly equivalent to:

```c
void __verona_sandbox_dispatcher(int fn, void *argframe) {
  switch(fn)
  {
    case {whatever the vtable index for std::vector<int>::pop_back is}:
       call_std_vector_int_pop(argrame);
       break;
    ....
  }
}
```

Eventually, the dispatcher function will expose an entire sequence of things in a `try` block as a single vtable entry, so we avoid multiple domain-transition round-trips but in the MVP any function that we call (including virtual functions) will be exposed this way (so we don't need to understand C++ vtables.  If `pop_back` were `virtual` then we'd generate exactly the same code.  This means that the only thing that we actually need are:

 - Call a function / method.
 - Reference fields in a structure.
 - Perform any necessary casts.

This is *far* simpler than a generic instruction builder.  The most complex part is likely to be the cast logic.

Note that this return function actually does the wrong thing: we will want it to store the return in a struct field and return nothing.

Reply:
So, IIUC, we'll need:
* Create structure types (either for Verona types mapped to C++ or internal compiler ones)
* Read / Write to their fields (this includes reading a pointer, which can be to a function)
* Direct / Indirect calls
* Simple control flow (switch / case)

To be clear, I didn't mean "a generic builder" like LLVM's `IRBuilder` (even though I have said _just that_ on one of the meetings), just some class separated from the `CXXInterface` that knows how to lower all the instructions we need (which is a small subset of all instructions possible).